### PR TITLE
Update VideoFrame ctor to take single buffer

### LINKED ISSUE #303: Update VideoFrame ctor to take single buffer
The original "planes" constructor was designed for coherence with the planes API. But Planes is now replaced by [frame.copyTo(destination, ...)](https://w3c.github.io/webcodecs/#dom-videoframe-copyto) (as of https://github.com/w3c/webcodecs/pull/208), which copies the entire frame copy to a single buffer. Its natural to also update the Planes constructor to match that style, taking in a single buffer and an optional sequence of PlaneLayouts. PR comments raised [the issue here](https://github.com/w3c/webcodecs/pull/208#issuecomment-836608874).

### PR DESCRIPTION
**Pull request by chcunningham**
The new single buffer replaces the planes ctor for better coherence (since frame.planes is now gone, replaced by frame.copyTo()). 

This also refactors a few frame.copyTo() algorithms to be re-usable between copyTo() and this new ctor() (both take a buffer, rect, and layout arguments, so they need the same validation). 

Fixes #303.

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jul 26, 2021, 10:37 PM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [CSS Spec Preprocessor](https://api.csswg.org/bikeshed/) - CSS Spec Preprocessor is the web service used to build Bikeshed specs.

:link: [Related URL](https://api.csswg.org/bikeshed/?url=https%3A%2F%2Fraw.githubusercontent.com%2Fw3c%2Fwebcodecs%2Fac786ff932365fe4863593c4bb975f36195d51a0%2Findex.src.html&md-warning=not%20ready)

```
<html><body><p><strong>WARNING: </strong>mysqli_connect(): (HY000/1040): Too many connections</p></body></html>
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/webcodecs%23304.)._
</details>

### CODE REVIEW

**Comment by chcunningham** [comment]
@sandersdan FYI

**Comment by chcunningham** [comment]
Editors call: @padenot taking closer look, but seems reasonable.

**Comment by chcunningham** [comment]
@padenot @aboba friendly ping

**Review by aboba — APPROVED** [review]

**Comment by chcunningham** [comment]
Editors call: @aboba lgtm

File: index.src.html [hunk 1]
```diff
@@ -2798,95 +2798,84 @@
...
+        |init|.{{VideoFrameBufferInit/codedHeight}},
+        |init|.{{VideoFrameBufferInit/format}}.
+5. If |parsedRect| is an exception, return |parsedRect|.
+6. Let |optLayout| be `undefined`.
+7. If |options|.{{VideoFrameBufferInit/layout}} [=map/exists=], assign its value
+    to |optLayout|.
+8. Let |combinedLayout| be the result of running the [=VideoFrame/Compute
+    Layout and Allocation Size=] algorithm with |parsedRect|,
+    |init|.{{VideoFrameBufferInit/format}}, and |optLayout|.
+9. If |combinedLayout| is an excpetion, throw |combinedLayout|.
```
Comment (padenot) [inline index.src.html:2823]: Typo

File: index.src.html [hunk 2]
```diff
@@ -2798,95 +2798,84 @@
...
-    ISSUE: The spec should define explicit rules for each
-        {{PixelFormat}} and reference them in the steps above. See
-        [#165](https://github.com/w3c/webcodecs/issues/165).
+1. If |init| is not a [=valid VideoFrameBufferInit=], throw a {{TypeError}}.
+2. Let |defaultRect| be «[ "x:" → `0`, "y" → `0`,
+    "width" → |init|.{{VideoFrameBufferInit/codedWidth}}, "height" →
+        |init|.{{VideoFrameBufferInit/codedWidth}} ]».
+3. Let |overrideRect| be `undefined`.
+4. If |init|.{{VideoFrameBufferInit/visibleRect}} [=map/exists=], assign its
+    value to overrideRect.
```
Comment (padenot) [inline index.src.html:2810]: Missing markup.

File: index.src.html [hunk 3]
```diff
@@ -2798,95 +2798,84 @@
...
-        |resource|.
-
-        ISSUE: The spec should provide definitions (and possibly diagrams) for
-            coded size, visible rectangle, and display size. See
-            [#166](https://github.com/w3c/webcodecs/issues/166).
-
-    8. Assign |resourceCodedWidth|, |resourceCodedHeight|,
+        size and plane strides to improve memory alignment. Increases will be
+        reflected by {{VideoFrame/codedWidth}} and {{VideoFrame/codedHeight}}.
+        Additionally, the User Agent may use
```
Comment (padenot) [inline index.src.html:2833]: User-Agent. Also, please don't use may in non-normative parts. This probably needs to be normative. Is there any reason this is not a normative requirement? Sounds like should to me maybe, that's often desirable but maybe nto always.

File: index.src.html (hunk 2 above)
Comment (chcunningham) [reply index.src.html]: Fixed.

File: index.src.html (hunk 1 above)
Comment (chcunningham) [reply index.src.html]: Fixed.

File: index.src.html (hunk 3 above)
Comment (chcunningham) [reply index.src.html:2949]: Fixed the note (made it normative). I was just confusing non-normative with normative-but-optional ("may"). 

Our existing mentions of "user agent" are without the hyphen. Looking at the html spec, I see this is far more common. Has the guidance changed to use a hyphen?

**Review by chcunningham — COMMENTED** [review]
Thanks @padenot

File: index.src.html (hunk 3 above)
Comment (padenot) [reply index.src.html:2949]: Not sure, and now I see that quite a few specs aren't in fact consistent. Let's not care too much.

**Comment by padenot** [comment]
Good for merging after rebase.

**Review by padenot — APPROVED** [review]

**Comment by chcunningham** [comment]
Thanks for reviews!

**Pull request merged**